sql/analyzer/symbol_resolution.go

```diff
@@ -0,0 +1,286 @@
+// Copyright 2020-2021 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package analyzer
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/plan"
+	"github.com/dolthub/go-mysql-server/sql/transform"
+)
+
+// findSubqueryExpr searches for a *plan.Subquery in a single node,
+// returning the subquery or nil
+func findSubqueryExpr(n sql.Node) *plan.Subquery {
+	var sq *plan.Subquery
+	ne, ok := n.(sql.Expressioner)
+	if !ok {
+		return nil
+	}
+	for _, e := range ne.Expressions() {
+		found := transform.InspectExpr(e, func(e sql.Expression) bool {
+			if e, ok := e.(*plan.Subquery); ok {
+				sq = e
+				return true
+			}
+			return false
+		})
+		if found {
+			return sq
+		}
+	}
+	return nil
+}
+
+// pruneTableCols uses a list of parent dependencies columns and stars
+// to prune the table schema
+func pruneTableCols(
+	n *plan.ResolvedTable,
+	needed map[tableCol]int,
```

Comment:
This type and name are pretty opaque

Recommend either a better name, a more complete comment, or a type alias that makes the meaning clear